exporter/elasticsearchexporter/model.go

```diff
@@ -169,123 +170,42 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
-		case pmetric.MetricTypeGauge:
-			dps = metric.Gauge().DataPoints()
-		case pmetric.MetricTypeSum:
-			dps = metric.Sum().DataPoints()
-		}
-
-		for j := 0; j < dps.Len(); j++ {
-			dp := dps.At(j)
-
-			hash := metricHash(dp.Timestamp(), dp.Attributes())
-			doc, docExists := docs[hash]
-			if !docExists {
-				doc = baseDoc.Clone()
-				doc.AddTimestamp("@timestamp", dp.Timestamp())
-				doc.AddAttributes("", dp.Attributes())
-
-				docs[hash] = doc
-			}
-
-			switch dp.ValueType() {
-			case pmetric.NumberDataPointValueTypeDouble:
-				doc.AddAttribute(metric.Name(), pcommon.NewValueDouble(dp.DoubleValue()))
-			case pmetric.NumberDataPointValueTypeInt:
-				doc.AddAttribute(metric.Name(), pcommon.NewValueInt(dp.IntValue()))
-			}
-		}
+func (m *encodeModel) encodeDocument(document objmodel.Document) ([]byte, error) {
+	if m.dedup {
+		document.Dedup()
+	} else if m.dedot {
+		document.Sort()
 	}
 
-	res := make([][]byte, 0, len(docs))
-
-	for _, doc := range docs {
-		if m.dedup {
-			doc.Dedup()
-		} else if m.dedot {
-			doc.Sort()
-		}
-
-		var buf bytes.Buffer
-		err := doc.Serialize(&buf, m.dedot)
-		if err != nil {
-			return nil, err
-		}
-
-		res = append(res, buf.Bytes())
+	var buf bytes.Buffer
+	err := document.Serialize(&buf, m.dedot)
+	if err != nil {
+		return nil, err
 	}
-
-	return res, nil
-}
-
-func metricHash(timestamp pcommon.Timestamp, attributes pcommon.Map) uint32 {
-	hasher := fnv.New32a()
-
-	timestampBuf := make([]byte, 8)
-	binary.LittleEndian.PutUint64(timestampBuf, uint64(timestamp))
-	hasher.Write(timestampBuf)
-
-	mapHash(hasher, attributes)
-
-	return hasher.Sum32()
+	return buf.Bytes(), nil
 }
 
-func mapHash(hasher hash.Hash, m pcommon.Map) {
-	m.Range(func(k string, v pcommon.Value) bool {
-		hasher.Write([]byte(k))
-		valueHash(hasher, v)
-
-		return true
-	})
-}
-
-func valueHash(h hash.Hash, v pcommon.Value) {
-	switch v.Type() {
-	case pcommon.ValueTypeEmpty:
-		h.Write([]byte{0})
-	case pcommon.ValueTypeStr:
-		h.Write([]byte(v.Str()))
-	case pcommon.ValueTypeBool:
-		if v.Bool() {
-			h.Write([]byte{1})
-		} else {
-			h.Write([]byte{0})
-		}
-	case pcommon.ValueTypeDouble:
-		buf := make([]byte, 8)
-		binary.LittleEndian.PutUint64(buf, math.Float64bits(v.Double()))
-		h.Write(buf)
-	case pcommon.ValueTypeInt:
-		buf := make([]byte, 8)
-		binary.LittleEndian.PutUint64(buf, uint64(v.Int()))
-		h.Write(buf)
-	case pcommon.ValueTypeBytes:
-		h.Write(v.Bytes().AsRaw())
-	case pcommon.ValueTypeMap:
-		mapHash(h, v.Map())
-	case pcommon.ValueTypeSlice:
-		sliceHash(h, v.Slice())
+func (m *encodeModel) upsertMetricDataPoint(documents map[uint32]objmodel.Document, resource pcommon.Resource, _ pcommon.InstrumentationScope, metric pmetric.Metric, dp pmetric.NumberDataPoint) error {
+	hash := metricHash(dp.Timestamp(), dp.Attributes())
+	var (
+		document objmodel.Document
+		ok       bool
+	)
+	if document, ok = documents[hash]; !ok {
+		document.AddAttributes("", resource.Attributes())
```

Comment:
Adding attributes to root means that this is ecs mode only. There is no support for none/raw mapping mode. Do we need them in this PR?

Reply:
Metrics status is "development", I suggest we leave this to a followup.